File: swim-js/swim-runtime-js/swim-core-js/@swim/codec/main/input/Input.ts
// Copyright 2015-2021 Swim Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {Lazy} from "@swim/util";
import {Mark} from "../source/Mark";
import {InputException} from "./InputException";
import {AnyInputSettings, InputSettings} from "./InputSettings";
import {InputEmpty} from "../"; // forward import
import {InputDone} from "../"; // forward import
import {InputError} from "../"; // forward import

/**
 * Non-blocking token stream reader, with single token lookahead.
 * `Input` enables incremental, interruptible parsing of network protocols and
 * data formats.
 *
 * ### Input tokens
 * Input tokens are modeled as primitive numbers, commonly representing Unicode
 * code points, or raw octets; each `Input` implementation specifies the
 * semantic type of its tokens.  The [[head]]  method peeks at the lookahead
 * token, without consuming it, and the [[step]] method advances the input to
 * the next token.
 *
 * ### Input states
 * An `Input` reader is always in one of three states: _cont_​inue, _empty_, or
 * _done_.  The _cont_ state indicates that a lookahead token is immediately
 * available; the _empty_ state indicates that no additional tokens are
 * available at this time, but that the stream may logically resume at some
 * point in the future; and the _done_ state indicates that the stream has
 * terminated.  [[isCont]] returns `true` when in the _cont_ state; [[isEmpty]]
 * returns `true` when in the _empty_ state; and [[isDone]] returns `true` when
 * in the _done_ state.
 *
 * ### Non-blocking behavior
 * `Input` readers never block.  An `Input` reader that would otherwise block
 * awaiting additional input instead enters the _empty_ state, signaling the
 * input consumer to back off processing the input, but to remain prepared to
 * process additional input in the future.  An `Input` reader enters the _done_
 * state when it encounters the final end of its input, signaling the input
 * consumer to stop processing.  [[Input.empty]] returns an `Input` reader in
 * the _empty_ state.  [[Input.done]] returns an `Input` reader in the _done_
 * state.
 *
 * ### Position tracking
 * The logical position of the lookahead token is made available via the
 * [[mark]] method, with optimized callouts for the byte [[offset]], one-based
 * [[line]] number, and one-based [[column]] in the current line.  The [[id]]
 * method returns a diagnostic identifier for the token stream.
 *
 * ### Cloning
 * An `Input` reader may be [[clone cloned]] to provide an indepently mutable
 * position into a shared token stream.  Not all `Input` implementations
 * support cloning.
 *
 * @see [[InputSettings]]
 * @see [[Parser]]
 */
export abstract class Input {
  /**
   * Returns `true` when a [[head lookahead]] token is immediately available.
   * i.e. this `Input` is in the _cont_ state.
   */
  abstract isCont(): boolean;

  /**
   * Returns `true` when no lookahead token is currently available, but
   * additional input may be available at some point in the future.  i.e. this
   * `Input` is in the _empty_ state.
   */
  abstract isEmpty(): boolean;

  /**
   * Returns `true` when no lookahead token is currently available, and no
   * additional input will ever become available.  i.e. this `Input` is in
   * the _done_ state.
   */
  abstract isDone(): boolean;

  /**
   * Returns `true` when no lookahead token is currently available due to an
   * error with the token stream. i.e. this `Input` is in the `error` state.
   * When `true`, `trap()` will return the input error
   */
  abstract isError(): boolean;

  /**
   * Returns `true` if this is a partial `Input` will that enter the _empty_
   * state after it consumes the last available input token.
   */
  abstract isPart(): boolean;

   /**
    * Returns a partial `Input` equivalent to this `Input`, if `part` is `true`;
    * returns a final `Input` equivalent to this `Input` if `part` is `false`.
    * The caller's reference to this `Input` should be replaced by the returned
    * `Input`
    */
  abstract asPart(part: boolean): Input;

  /**
   * Returns the current lookahead token, if this `Input` is in the
   * _cont_ state.
   *
   * @throws [[InputException]] if this `Input` is not in the _cont_ state.
   */
  abstract head(): number;

  /**
   * Advances to the next token, if this `Input` is in the _cont_ state.
   *
   * @throws `Error` if this `Input` is not in the _cont_ state.
   */
  abstract step(): Input;

  /**
   * Sets the position of this `Input` to the given `mark`.  Rewinds to the
   * start of this `Input`, if `mark` is `undefined`.
   *
   * @throws [[InputException]] if this `Input` does not support seeking,
   *         or is unable to reposition to the `mark`.
   */
  abstract seek(mark?: Mark): Input;

  /**
   * Returns the input error when in the _error_ state
   *
   * @throws InputException if the `Input` is not the _error_state
   */
  trap(): Error {
    throw new InputException();
  }

  /**
   * An informative identifier for this token stream, or `undefined` if this
   * stream is unidentified.
   */
  abstract readonly id: string | undefined;

  /**
   * Returns an `Input` equivalent to this `Input`, but logically identified
   * by the given–possibly `undefined`–`id`.  The caller's reference to this
   * `Input` should be replaced by the returned `Input`.
   */
  abstract withId(id: string | undefined): Input;

  /**
   * The position of the current lookahead token, relative to the start
   * of the stream.
   */
  abstract readonly mark: Mark;

  /**
   * Returns an `Input` equivalent to this `Input`, but logically positioned
   * at the given `mark`.  The physical position in the input stream is not
   * modified.  The caller's reference to this `Input` should be replaced by
   * the returned `Input`.
   */
  abstract withMark(mark: Mark): Input;

  /**
   * The byte offset of the current lookahead token, relative to the start
   * of the stream.
   */
  abstract readonly offset: number;

  /**
   * The one-based line number of the current lookahead token, relative to
   * the start of the stream.
   */
  abstract readonly line: number;

  /**
   * The one-based column number of the current lookahead token, relative to
   * the current line in the stream.
   */
  abstract readonly column: number;

  /**
   * The `InputSettings` used to configure the behavior of input consumers
   * that read from this `Input`.
   */
  abstract readonly settings: InputSettings;

  /**
   * Returns a clone of this `Input` with the given `settings`.
   *
   * @throws `Error` if this `Input` reader cannot be cloned.
   */
  abstract withSettings(settings: AnyInputSettings): Input;

  /**
   * Returns an independently positioned view into the token stream,
   * initialized with identical state to this `Input`.
   *
   * @throws `Error` if this `Input` reader cannot be cloned.
   */
  abstract clone(): Input;

  /**
   * Returns an `Input` reader in the _empty_ state.
   */
  @Lazy
  static empty(): Input {
    return new InputEmpty(void 0, Mark.zero, InputSettings.standard());
  }

  /**
   * Returns an `Input` reader in the _done_ state.
   */
  @Lazy
  static done(): Input {
    return new InputDone(void 0, Mark.zero, InputSettings.standard());
  }

  /**
   * Returns an `Input` in the _error_ state that traps the given `error`.
   */
  static error(error: Error): Input {
    return new InputError(error, void 0, Mark.zero, InputSettings.standard());
  }
}

File: swim-java/swim-core/swim.codec/src/main/java/swim/codec/Input.java
// Copyright 2015-2022 Swim.inc
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.codec;

import swim.annotations.CheckReturnValue;
import swim.annotations.Nullable;
import swim.annotations.Public;
import swim.annotations.Since;

/**
 * A non-blocking chunked input stream. {@code Input} facilitates
 * interruptible parsing of network protocols and data formats.
 *
 * <h2>Input tokens</h2>
 * <p>
 * Input tokens are modeled as primitive {@code int} values, commonly
 * representing Unicode code points, or raw octets. The semantics of
 * input tokens is specified by individual {@code Input} subclasses.
 * <p>
 * The {@link #head()} method peeks at the current lookahead token,
 * without consuming it. The {@link #step()} method advances the input
 * to the next state.
 *
 * <h2>Input states</h2>
 * <p>
 * An {@code Input} instance is always in one—and only one—of the following
 * four <em>input states</em>:
 * <ul>
 * <li>{@link #isCont() input-cont}: an input token is available
 *     at the {@link #head()} of the input stream
 * <li>{@link #isEmpty() input-empty}: the input stream has reached
 *     the end of the current input chunk
 * <li>{@link #isDone() input-done}: the input stream has reached
 *     the end of the last input chunk
 * <li>{@link #isError() input-error}: the input stream terminated
 *     with an {@linkplain #getError() input error}
 * </ul>
 *
 * <h2>Input chunks</h2>
 * <p>
 * {@code Input} methods never block. An input stream that would otherwise
 * need to block awaiting additional input instead enters the {@code input-empty}
 * state, signaling the input consumer to asynchronously wait for additional
 * input before continuing. If no additional input will ever become available,
 * the input stream enters the {@code input-done} state, signaling the input
 * consumer to terminate. The {@link #isLast()} method indicates whether the
 * input stream will enter the {@code input-empty} state or the {@code input-done}
 * state when it reaches the end of the current input chunk.
 *
 * <h2>Position tracking</h2>
 * <p>
 * The logical location of the current lookahead token is made available via
 * the {@link #location()} method, with convenience accessors for the byte
 * {@linkplain #offset() offset}, one-based {@linkplain #line() line} number,
 * and one-based {@linkplain #column() column} in the current line. The
 * {@link #name()} method returns an optional name for the input source.
 *
 * <h2>Cloning</h2>
 * <p>
 * An input stream may be {@linkplain #clone() cloned} to provide an
 * independently mutable position into a shared input chunk. Not all
 * {@code Input} implementations support cloning.
 *
 * @see Parser
 * @see InputBuffer
 */
@Public
@Since("5.0")
public abstract class Input {

  /**
   * Constructs a new {@code Input} instance.
   */
  protected Input() {
    // nop
  }

  /**
   * Returns {@code true} when in the {@code input-cont} state, with an
   * input token available at the {@link #head()} of the input stream
   *
   * @return whether or not this {@code Input} instance is
   *         in the {@code input-cont} state
   */
  public abstract boolean isCont();

  /**
   * Returns {@code true} when in the {@code input-done} state,
   * having reached the end of the current input chunk.
   *
   * @return whether or not this {@code Input} instance is
   *         in the {@code input-done} state
   */
  public abstract boolean isEmpty();

  /**
   * Returns {@code true} when in the {@code input-done} state,
   * having reached the end of the last input chunk.
   *
   * @return whether or not this {@code Input} instance is
   *         in the {@code input-done} state
   */
  public abstract boolean isDone();

  /**
   * Returns {@code true} when in the {@code input-error} state,
   * having terminated with an {@linkplain #getError() input error}.
   *
   * @return whether or not this {@code Input} instance is
   *         in the {@code input-error} state
   */
  public abstract boolean isError();

  /**
   * Returns {@code true} when in either the {@code input-cont} state
   * or the {@code input-done} state, indicating that parsing can
   * potentially make further progress.
   *
   * @return whether or not this {@code Input} instance is in either the
   *         {@code input-cont} state or the {@code input-done} state
   */
  public boolean isReady() {
    return this.isCont() || this.isDone();
  }

  /**
   * Returns {@code true} when in either the {@code input-empty} state
   * or the {@code input-error} state, indicating that parsing can't
   * make further progress at this time.
   *
   * @return whether or not this {@code Input} instance is in either the
   *         {@code input-empty} state or the {@code input-error} state
   */
  public boolean isBreak() {
    return this.isEmpty() || this.isError();
  }

  /**
   * Returns {@code true} if this input stream will enter the
   * {@code input-done} state when it reaches the end of the current
   * input chunk; otherwise returns {@code false} if the input stream
   * will enter the {@code input-empty} state when it reaches the end
   * of the current input chunk.
   *
   * @return whether this input stream will enter the {@code input-done}
   *         state or the {@code input-empty} state when it reaches the end
   *         of the current input chunk
   */
  public abstract boolean isLast();

  /**
   * Sets the {@link #isLast()} flag to {@code last} and returns {@code this}.
   * If {@code last} is {@code true}, the input stream will enter the
   * {@code input-done} state when it reaches the end of the current
   * input chunk; if {@code last} is {@code false}, the input stream
   * will enter the {@code input-empty} state when it reaches the end
   * of the current input chunk.
   *
   * @param last whether the input stream should enter the {@code input-done}
   *        state or the {@code input-empty} state when it reaches the end
   *        of the current input chunk
   * @return {@code this}
   */
  public abstract Input asLast(boolean last);

  /**
   * Returns the current lookahead token, if in the {@code input-cont} state.
   *
   * @return the current lookahead token, if available
   * @throws IllegalStateException if not in the {@code input-cont} state
   */
  public abstract int head();

  /**
   * Returns the {@code k}-th lookahead token, if available; otherwise returns
   * {@code -1} if fewer than {@code k + 1} lookahead tokens are available in
   * the current input chunk, or if the input stream doesn't support
   * multi-token lookahead.
   *
   * @param k the offset from the current position in the input chunk
   *        of the token to return
   * @return the input token {@code k} steps forward in the input chunk,
   *         or {@code -1} if the lookahead token is not available
   */
  public abstract int lookahead(int k);

  /**
   * Consumes the current lookahead token, advances the input stream
   * to the next state, and returns {@code this}.
   *
   * @return {@code this}
   * @throws IllegalStateException if not in the {@code input-cont} state
   */
  public abstract Input step();

  /**
   * Sets the current position of this {@code Input} instance to the given
   * {@code position} in the input stream and returns {@code this}.
   *
   * @param position the position in the input stream from which
   *        input tokens should be consumed
   * @return {@code this}
   * @throws UnsupportedOperationException if the input stream
   *         does not support seeking
   * @throws IllegalArgumentException when attempting to seek
   *         to an invalid source {@code position}
   */
  public abstract Input seek(@Nullable SourcePosition position);

  /**
   * Returns the logical location of the current lookahead token
   * in the input stream.
   *
   * @return the source position of the current lookahead token
   *         in the input stream
   */
  public abstract SourcePosition location();

  /**
   * Sets the logical location of the current lookahead token to the given
   * {@code location} in the input stream and returns {@code this}.
   * This method does not seek the stream; it's used to offset
   * source positions of continued input chunks.
   *
   * @param location the source position of the current lookahead token
   *        in the input stream
   * @return {@code this}
   */
  public abstract Input location(SourcePosition location);

  /**
   * Returns a name for the source of this input stream,
   * or {@code null} if the input source is unnamed.
   *
   * @return the name of the source of this input stream
   */
  public @Nullable String name() {
    return this.location().name();
  }

  /**
   * Sets the name of the source of this input stream and returns {@code this}.
   *
   * @param name the name to use for the source of this input stream
   * @return {@code this}
   */
  public abstract Input name(@Nullable String name);

  /**
   * Returns the byte offset of the current lookahead token
   * relative to the start of the input stream.
   *
   * @return the byte offset of the current location in the input stream
   */
  public long offset() {
    return this.location().offset();
  }

  /**
   * Returns the one-based line number of the current lookahead token
   * relative to the start of the input stream.
   *
   * @return the line number of the current location in the input stream
   */
  public int line() {
    return this.location().line();
  }

  /**
   * Returns the one-based column number of the current lookahead token
   * relative to the current line in the input stream.
   *
   * @return the column number of the current location in the input stream
   */
  public int column() {
    return this.location().column();
  }

  /**
   * Returns the input error, if in the {@code input-error} state;
   * otherwise throws an unchecked exception.
   *
   * @return the input error, if present
   * @throws IllegalStateException if not in the {@code input-error} state
   */
  @CheckReturnValue
  public Throwable getError() {
    throw new IllegalStateException("no input error");
  }

  /**
   * Returns an independently positioned view of the input stream,
   * initialized with identical state to this {@code Input} instance.
   *
   * @return a new {@code Input} instance that is independent
   *         of this input stream
   * @throws UnsupportedOperationException if this input stream
   *         can't be cloned
   */
  @Override
  public abstract Input clone();

}
